### DISCUSSION TITLE
Problem with a smart Fan

### DISCUSSION THREAD

**Discussion by G-D-D-971** [HomeSpan/HomeSpan#815] [Q&A] [answered]
Hello, 
I am very interested in this completely unique and fun library, all the devices in my home are connected to HomeKit, but the small problem is, that I really want to create my own accessories. The problem, I am a novice and I admit that despite my experience with the arduino language and the arduino IDE I have some difficulty creating my own fan that also has a temperature and humidity sensor. 
So that's my problem, although I managed to create the code for the temperature sensors largely inspired by a tutorial, that of fan management is much more complex, when I solve a problem, another arises, I used the same structure as the example file "6 - Dimmable LED", but despite my modifications nothing helps... 
If someone could help me, I did not find any examples on the internet. 
The thing is that I don't know if what I want to do is possible, it's that I want my fan and sensors to be in the same accessories (I attach a photograph of a HomePod from my home app to illustrate). 

**Captures :**
<img width="500" alt="Homepod_accessories-1" src="https://github.com/HomeSpan/HomeSpan/assets/124715754/1350d7f5-7277-4185-985e-0ecaab75c749">
<img width="506" alt="Homepod_accessories-2" src="https://github.com/HomeSpan/HomeSpan/assets/124715754/4eba4469-dbd7-411d-9d4f-fe47592efc6e">

**My Arduino code :**
```
#include "HomeSpan.h"         // Always start by including the HomeSpan library
#include "DEV_Identify.h"      
#include "DEV_Sensors.h"

void setup() {
  homeSpan.setControlPin(5);
  homeSpan.setStatusPin(4);
  homeSpan.setPairingCode("00000000");
  homeSpan.setQRID("1111");
  Serial.begin(115200); 

  homeSpan.begin(Category::Fans,"VMC");

  new SpanAccessory();
    new Service::AccessoryInformation();                
      new Characteristic::Identify();                         

  new SpanAccessory();
    new DEV_Fan();                                                        
    new DEV_TempSensor();
    new DEV_HumSensor();

} // end of setup()

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();         // run HomeSpan!
  
} // end of loop()
```

**My DEV_Identify.h :**
```
struct DEV_Identify : Service::AccessoryInformation {

   int nBlinks;                    // number of times to blink built-in LED in identify routine
   SpanCharacteristic *identify;   // reference to the Identify Characteristic
   
   DEV_Identify(const char *name, const char *manu, const char *sn, const char *model, const char *version, int nBlinks) : Service::AccessoryInformation(){
     
     new Characteristic::Name("VMC");                   // create all the required Characteristics with values set based on above arguments
     new Characteristic::Manufacturer("SocieteNS");
     new Characteristic::SerialNumber("971412024");    
     new Characteristic::Model("VMC-1/A");
     new Characteristic::FirmwareRevision("1.0.0");
     new Characteristic::HardwareRevision("1.0.0");
     identify=new Characteristic::Identify();          // store a reference to the Identify Characteristic for use below

     this->nBlinks=nBlinks;                            // store the number of times to blink the FAN

     pinMode(homeSpan.getStatusPin(),OUTPUT);          // make sure FAN is set for output
}

   boolean update(){
       
     for(int i=0;i<nBlinks;i++){
       digitalWrite(homeSpan.getStatusPin(),LOW);
       delay(250);
       digitalWrite(homeSpan.getStatusPin(),HIGH);
       delay(250);
    }

     return(true);                               // return true
     
  } // update
   
 };
```

**My DEV_Sensors.h:**
```
 #include "DHT.h"

 #define DHTPIN 17
 #define DHTTYPE DHT22

 // reference to the Sensor Objects
 DHT dht(DHTPIN, DHTTYPE);

 // A standalone Temperature sensor
 struct DEV_TempSensor : Service::TemperatureSensor {    

   // reference to the Current Temperature Characteristic
   SpanCharacteristic *temp;                                      

   // constructor() method
   DEV_TempSensor() : Service::TemperatureSensor() {      

     // start dhttemp Object
     dht.begin();                                    

     // instantiate the Current Temperature Characteristic
         temp = new Characteristic::CurrentTemperature(-10.0);    
     // expand the range from the HAP default of 0-100 to -50 to 100 to allow for negative temperatures
     temp->setRange(-50, 100);                                

     // initialization message
     Serial.print("Configuring Temperature Sensor");          
     Serial.print("\n");

  } // end constructor

   void loop() {

     // the temperature refreshes every 20 seconds Par the elapsed time
     if (temp->timeVal() > 20000) {
       // read temperature from sensor dht22
       float temperature = dht.readTemperature();        
       // set the new temperature; this generates an Event Notification and also resets the elapsed time
       temp->setVal(temperature);                            

       LOG1("Temperature Update: ");
       LOG1(temperature);
       LOG1(" ; ");
    }
  } // loop
 };

 // A standalone Humidity sensor
 struct DEV_HumSensor : Service::HumiditySensor {    

   // reference to the Current Humidity Characteristic
   SpanCharacteristic *hum;                                  

   // constructor() method
   DEV_HumSensor() : Service::HumiditySensor() {
     // start dhthum Object
     dht.begin();                                  

     // instantiate the Current Temperature Characteristic
     hum = new Characteristic::CurrentRelativeHumidity(50);
     // expand the range to 20%-100%
     hum->setRange(20, 100);                                

     // initialization message
     Serial.print("Configuring Humidity Sensor");          
     Serial.print("\n");

  } // end constructor

   void loop() {
     // the humidity refreshes every 20 seconds Par the elapsed time
         if (hum->timeVal() > 20000) {
       // read humidity from sensor dht22
       float humidity = dht.readHumidity();  
       // set the new humidity; this generates an Event Notification and also resets the elapsed time        
       hum->setVal(humidity);                            
       
       LOG1("Humidity Update: ");
       LOG1(humidity);
       LOG1(" ; ");
    }
  } // loop
 };

 struct DEV_Fan : Service::Fan {      // Advenced Fan

  FanPin *fanPin;                                   // NEW! Create reference to Fan Pin instantiated below
  SpanCharacteristic *power;                        // reference to the On Characteristic
  SpanCharacteristic *speed;                        // NEW! Create a reference to the Brightness Characteristic instantiated below
  
  DEV_Fan(int pin) : Service::Fan(){       // constructor() method

    power=new Characteristic::Active();     
                
    speed=new Characteristic::RotationSpeed(50);       // NEW! Instantiate the Brightness Characteristic with an initial value of 50% (same as we did in Example 4)
    speed->setRange(0,100,25);                       // NEW! This sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1% (different from Example 4 values)

    this->fanPin=new FanPin(pin);                   // NEW! Configures a PWM FAN for output to the specified pin.  Note pinMode() does NOT need to be called in advance
    
  } // end constructor

  boolean update(){                              // update() method

    // Here we set the brightness of the LED by calling ledPin->set(brightness), where brightness=0-100.
    // Note HomeKit sets the on/off status of a LightBulb separately from its brightness, which means HomeKit
    // can request a LightBulb be turned off, but still retains the brightness level so that it does not need
    // to be resent once the LightBulb is turned back on.
    
    // Multiplying the newValue of the On Characteristic ("power", which is a boolean) with the newValue of the
    // Brightness Characteristic ("level", which is an integer) is a short-hand way of creating the logic to
    // set the LED level to zero when the LightBulb is off, or to the current brightness level when it is on.
    
    fanPin->set(power->getNewVal()*speed->getNewVal());
   
    return(true);                               // return true
  
  } // update
```

Thanks for all of your help,
Nathan.

**Comment by HomeSpan** [comment]
The Home App always shows temperature, humidity (and other sensor readings) as standalone tiles grouped by sensor type.  In other words, if you had 5 Accessories each defining a Temperature Sensor, all the temperature sensors would be grouped by the Home App under the temperature sensor tile even though controls for those Accessories would each be shown on a separate Tile.

This was a design change Apple made quite a while ago (maybe iOS 14?) and unfortunately there is not much that can be done to modify the approach and show everything together under the controlling Accessory Tile.  The same is true even for commercial devices.

One note: the single-wire DHT sensor you seem to be using is inherently unstable in terms of connectivity.  You will likely have problems getting that to work reliably with HomeSpan (many shops have stopped selling them and replaced them instead with similar models driven by the industry-standards I2C protocol, such as the DHT20,  which works very well with HomeSpan).

**Reply by G-D-D-971** [reply]
The DHT22 sensor uses I2C communication well, the one I will order has 4 pins, 3 of which are used in the project.
Do you have examples of code to operate an engine in PWM?
In addition, combined accessories are supported by commercial devices as evidenced by the screenshots I attach to you of one of my cameras that have a motion sensor and an integrated lamp. It seems to me that it also works on HomeSpan, the example of the fan with the integrated lamp appears as a secondary accessory It seems to me...
<img width="506" alt="Capture d’écran, le 2024-03-23 à 13 47 09" src="https://github.com/HomeSpan/HomeSpan/assets/124715754/15891fb7-cecc-4738-960b-b9c0a4cf33d8">
<img width="493" alt="Capture d’écran, le 2024-03-23 à 13 46 53" src="https://github.com/HomeSpan/HomeSpan/assets/124715754/9a6960f6-265b-4142-b05b-c3532ad9e6e6">

**Reply by HomeSpan** [reply]
Are you sure you don't have a DHT20?  The specs on the web show the 20 as I2C and the 22 as single wire. 

As long as all your services are under a single Accessory they are combined.  But the Home App chooses how to render them.

**Reply by G-D-D-971** [reply]
Okay very well, I thought there was a specific option for the combination 😅.
Indeed, I had not been careful about the choice of the sensor, I will follow your advice and take the DHT20, I just wanted to follow [the tutorial they proposed](https://www.az-delivery.de/fr/blogs/azdelivery-blog-fur-arduino-und-raspberry-pi/apple-homekit-mit-esp32-teil-2), not knowing how to use the library completely (which is great 🤩).

Could you give me a functional code for the control of a PWM motor and for that of the DHT20 sensor? Thank you in advance.

**Comment by G-D-D-971** [comment]
Would a person have code for controlling an motor in PWM and retrieving the values of a DHT20 on homespan? 
I understand the code when I have it in front of me but otherwise... I don't know how to write it at all.

**Reply by G-D-D-971** [reply, marked as answer]
Hello, for those like me who are looking to control a non-servomotor with PWM but who have not succeeded here is the code I use, it has not been checked and I do not know if it works...
```
struct DEV_Fan : Service::Fan {       // Dimmable Fan

  LedPin *ledPin;                                   // NEW! Create reference to Fan Pin instantiated below
  SpanCharacteristic *power;                        // reference to the On Characteristic
  SpanCharacteristic *level;                        // NEW! Create a reference to the speed Characteristic instantiated below
  
  DEV_Fan(int pin) : Service::Fan(){       // constructor() method

    power=new Characteristic::Active();     
                
    level=new Characteristic::RotationSpeed(50);       // NEW! Instantiate the speed Characteristic with an initial value of 50% (same as we did in Example 4)
    level->setRange(0,100,25);                       // NEW! This sets the range of the speed to be from a min of 0%, to a max of 100%, in steps of 25% (different from Example 4 values)

    this->ledPin=new LedPin(pin);                   // NEW! Configures a PWM motor for output to the specified pin.  Note pinMode() does NOT need to be called in advance
    
  } // end constructor

  boolean update(){                              // update() method
    
    ledPin->set(power->getNewVal()*level->getNewVal());    
   
    return(true);                               // return true
  
  } // update
};
```
I will test it as soon as possible...
[Update] The code work perfectly…

**Reply by alexandrumanta13** [reply]
I have uploaded your code and when i want to pair i receive this message: Accessory is out of compliance :(

**Reply by G-D-D-971** [reply]
Hi @alexandrumanta13, my code is just to control an engine with PWM on homeSpan, you are supposed to install this code in a separate file and call it in the Homespan main file with a function such as:
```
  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
    new DEV_Fan(15);
```
Where (15) obviously refers to the engine pin.
I myself tested the code in real condition and it is part of a larger project where it is the central part.

**Reply by alexandrumanta13** [reply]
Hi @G-D-D-971! I've made some changes and now is working! Thank you!

**Reply by G-D-D-971** [reply]
You'r Welcome !

**Comment by wingspinner** [comment]
> I understand the code when I have it in front of me but otherwise... I don't know how to write it at all.


ChatGPT might be able to help you write the code and learn at the same time the latest version can actually write some decent code if you give it enough info.

**Reply by G-D-D-971** [reply]
Yes ChatGPT can be useful but we must not forget that it is a chatbot that generates text (and code) based on probabilities, he is not aware of what he writes or even if it is correct. In addition, HomeSpan is a minor library in the extremely vast universe of Arduino librairies(Arduino who is himself very small compared to other constructor/language) so it is unlikely that ChatGPT does not know HomeSpan well or badly or even that he invents code... you must always look and understand what it gives. having a base in code is the essential, but I agree that it is rather good...
